event/src/main/java/com/bazaarvoice/emodb/event/db/astyanax/ChannelAllocationState.java

```diff
@@ -53,22 +59,45 @@ public synchronized SlabRef detach() {
         return slab;
     }
 
-    public synchronized SlabAllocation allocate(int desired) {
+    public synchronized SlabAllocation allocate(int desired, Collection<Integer> eventSizes) {
```

Comment:
Based on our conversation offline on the need to maintain and index in the `eventSizes` collection as we drain it, I created a base implementation for allocate that may be simpler. Of course, it is a method with side effects, but I'm not sure what we can cleanly do about that without changing the whole paradigm.

```
public synchronized SlabAllocation allocate(PeekingIterator<Integer> eventSizes) {
        checkArgument(eventSizes.hasNext());

        if (!isAttached()) {
            return null;
        }

        int offset = _slabConsumed;
        int remaining = Constants.MAX_SLAB_SIZE - _slabConsumed;
        long remainingBytes = Constants.MAX_SLAB_SIZE_IN_BYTES - _slabBytesConsumed;
        int eventsInserted = 0;

        while (remaining > 0) {
            if (eventSizes.hasNext()) {
                int eventBytes = eventSizes.peek();
                if ((remainingBytes -= eventBytes) >= 0) {
                    remaining--;
                    _slabConsumed++;
                    _slabBytesConsumed += eventBytes;
                    eventsInserted++;
                    eventSizes.next();
                } else {
                    break;
                }
            } else {
                // All events fit in single slab.
                return new DefaultSlabAllocation(_slab.addRef(), offset, eventsInserted);
            }
        }

        // Not every event could fit in the currently attached slab. Return the rest of the slab. Detach from it so we'll allocate a new one next time.
        return new DefaultSlabAllocation(detach(), offset, eventsInserted);
    }
```

Reply:
OK